front/lib/api/vaults.ts

```diff
@@ -148,3 +150,62 @@ export async function hardDeleteVault(
 
   return new Ok(undefined);
 }
+
+export async function updateVaultPermissions(
+  auth: Authenticator,
+  vault: VaultResource,
+  {
+    isRestricted,
+    memberIds,
+  }: { isRestricted: boolean; memberIds: string[] | null }
+): Promise<Result<undefined, DustError | Error>> {
+  if (!vault.canAdministrate(auth)) {
+    return new Err(
+      new DustError("unauthorized", "Cannot update permissions for vault.")
+    );
+  }
+
+  const regularGroups = vault.groups.filter(
+    (group) => group.kind === "regular"
+  );
+  // Assert that there is exactly one regular group associated with the vault.
+  assert(
+    regularGroups.length === 1,
+    `Expected exactly one regular group for the vault, but found ${regularGroups.length}.`
+  );
+  const [defaultVaultGroup] = regularGroups;
+
+  const wasRestricted = vault.groups.every((g) => !g.isGlobal());
+
+  if (isRestricted) {
+    // If the vault should be restricted and was not restricted before, remove the global group.
+    if (!wasRestricted) {
+      const updateRes = await vault.updatePermissions(auth, true);
+      if (updateRes.isErr()) {
+        return updateRes;
+      }
+    }
+
+    if (memberIds) {
+      const users = await UserResource.fetchByIds(memberIds);
+
+      return defaultVaultGroup.setMembers(
+        auth,
+        users.map((u) => u.toJSON())
+      );
+    }
+
+    return new Ok(undefined);
+  } else {
+    // If the vault should not be restricted and was restricted before, add the global group.
+    if (wasRestricted) {
+      const updateRes = await vault.updatePermissions(auth, false);
+      if (updateRes.isErr()) {
+        return updateRes;
+      }
+    }
+
+    // Remove all members.
+    return defaultVaultGroup.setMembers(auth, []);
```

Comment:
Are you sure about that ?
I would personally keep the members.

Reply:
What's the value of keeping the members? Given that today, those members don't get another role, I personally find it easier to read it as a big if, either the vault is not restricted, has a global vault or has members.

Tomorrow it might make sense though, if we start assigning a role to a group.

Reply:
I think I'll be sad if I want to temporaly share a vault, then go back to restricted and see all my members gone.